### PR TITLE
inverse kinematics works with rotation constraints

### PR DESCRIPTION
**Pull request by AndrewMeadows**
This PR fixes some bugs when using IK with constraints at the joints.

Also modifies how the hand targets are rendered -- the XYZ axes of the targets are rendered and the "finger tip" is shown as a green sphere on the y-axis.

### CODE REVIEW

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 1]
```diff
@@ -203,7 +203,13 @@ const AnimPoseVec& AnimInverseKinematics::evaluate(const AnimVariantMap& animVar
...
                                 if (constraint) {
-                                    constraint->apply(newRot);
+                                    bool constrained = constraint->apply(newRot);
+                                    if (constrained) {
+                                        // the constraint will modify the movement of the tip so we have to compute the modified
+                                        // model-frame deltaRotation
+                                        // Q' = Qp^ * dQ * Q  -->  dQ =   Qp * Q' * Q^
+                                        deltaRotation = absolutePoses[parentIndex].rot * newRot * glm::inverse(absolutePoses[index].rot);
+                                    }
                                 }
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: This was an important fix: when the constraint modifies the rotation we must compute the effective delta-rotation in the model frame so we can correctly update the "tip" position as we iterate down (up?) the skeleton joints.

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 2]
```diff
@@ -258,7 +264,18 @@ const AnimPoseVec& AnimInverseKinematics::evaluate(const AnimVariantMap& animVar
...
+        // relax toward underpose
+        const float RELAXATION_TIMESCALE = 0.25f;
+        const float alpha = glm::clamp(dt / RELAXATION_TIMESCALE, 0.0f, 1.0f);
+        int numJoints = (int)_relativePoses.size();
+        for (int i = 0; i < numJoints; ++i) {
+            float dotSign = copysignf(1.0f, glm::dot(_relativePoses[i].rot, _defaultRelativePoses[i].rot));
+            _relativePoses[i].rot = glm::normalize(glm::lerp(_relativePoses[i].rot, dotSign * underPoses[i].rot, alpha));
+        }
+    }
     return evaluate(animVars, dt, triggersOut);
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: The IK is much more stable if we always start at the last solution but relax toward the underpose (e.g. the target animation state before IK).

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 3]
```diff
@@ -558,14 +604,11 @@ void AnimInverseKinematics::setSkeletonInternal(AnimSkeleton::ConstPointer skele
...
 
-    // No constraints!
-    /*
     if (skeleton) {
         initConstraints();
     } else {
         clearConstraints();
     }
-    */
 }
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: We now create some default constraints for the arm and spine joints.

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 4]
```diff
@@ -516,7 +562,7 @@ void AnimInverseKinematics::initConstraints() {
             eConstraint->setAngleLimits(minAngle, maxAngle);
 
             constraint = static_cast<RotationConstraint*>(eConstraint);
-        } else if (0 == name.compare("Foot", Qt::CaseInsensitive)) {
+        } else if (0 == baseName.compare("FootXXX", Qt::CaseInsensitive)) {
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: The "XXX" is to disable default constraints for these joints until later testing.

File: libraries/animation/src/SwingTwistConstraint.cpp [hunk 5]
```diff
@@ -201,6 +241,7 @@ bool SwingTwistConstraint::apply(glm::quat& rotation) const {
     }
 
     if (swingWasClamped || twistWasClamped) {
+        // update the rotation
         twistRotation = glm::angleAxis(clampedTwistAngle, yAxis);
         rotation = swingRotation * twistRotation * _referenceRotation;
         return true;
```
Comment (AndrewMeadows) [inline libraries/animation/src/SwingTwistConstraint.cpp]: These changes reduce the "popping" when greatly exceeding twist limits and rolling around toward the far limits on the other side.

File: libraries/animation/src/AnimInverseKinematics.cpp [hunk 6]
```diff
@@ -422,15 +455,28 @@ void AnimInverseKinematics::initConstraints() {
...
             stConstraint->setSwingLimits(swungDirections);
+            */
 
-            /*
+            // simple cone
             std::vector<float> minDots;
-            const float MAX_HAND_SWING = PI / 3.0f;
+            const float MAX_HAND_SWING = PI / 2.0f;
             minDots.push_back(cosf(MAX_HAND_SWING));
             stConstraint->setSwingLimits(minDots);
```
Comment (AndrewMeadows) [inline libraries/animation/src/AnimInverseKinematics.cpp]: Although PI/2 is a bit far for the wrist to bend, especially about its z-axis (normal to palm), this simple cone limit works well enough.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c589f0e58e2a01a62d1404ea321b4dcce9199eb6 available at: https://gist.github.com/5f29d6ddf89ad4a3d799<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5786.dmg

**Comment by AndrewMeadows** [comment]
Don't merge -- there is a crash bug I'm hunting.

**Comment by AndrewMeadows** [comment]
Ok, this is good to go.

I've noticed that my FPS has gone way down after adding the IK computations (taking up to 8 msec to iterate 16 times over three targets) for the debug build.  This makes me sad, however these calculations are 10x faster in an optimized release build (on linux).

**Comment by hifi-gustavo** [comment]
Build console output for commit: d01c5d1ae5ea9102af84165c9690a07ba52991a7 available at: https://gist.github.com/5f29d6ddf89ad4a3d799<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5786.dmg

File: interface/src/avatar/Hand.cpp [hunk 7]
```diff
@@ -40,76 +39,64 @@ void Hand::simulate(float deltaTime, bool isMine) {
     }
 }
 
-void Hand::render(RenderArgs* renderArgs, bool isMine) {
+void Hand::renderHandTargets(RenderArgs* renderArgs, bool isMine) {
+    const float avatarScale = DependencyManager::get<AvatarManager>()->getMyAvatar()->getScale();
```
Comment (howard-stearns) [inline interface/src/avatar/Hand.cpp]: _owningAvatar->getScale() instead?  (Avoid use of DependencyManger pseudo-globals, which will cause problems for threading.)

**Comment by hifi-gustavo** [comment]
Build console output for commit: e85cdad8c7a2520c7b6a53895690fb8994c41ce3 available at: https://gist.github.com/5f29d6ddf89ad4a3d799<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5786.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5786.dmg

**Pull request merged**